packages/@aws-cdk/cdk/lib/app.ts

```diff
@@ -34,73 +33,49 @@ export class App extends Root {
   /**
    * Runs the program. Output is written to output directory as specified in the request.
    */
-  public run(): void {
+  public run(): ISynthesisSession {
+    // this app has already been executed, no-op for you
+    if (this._session) {
+      return this._session;
+    }
+
     const outdir = process.env[cxapi.OUTDIR_ENV];
-    if (!outdir) {
-      process.stderr.write(`ERROR: The environment variable "${cxapi.OUTDIR_ENV}" is not defined\n`);
-      process.stderr.write('AWS CDK Toolkit (>= 0.11.0) is required in order to interact with this program.\n');
-      process.exit(1);
-      return;
+    if (outdir) {
+      this._session = new SynthesisSession({ outdir });
+    } else {
+      this._session = new InMemorySynthesisSession();
     }
 
-    const result: cxapi.SynthesizeResponse = {
-      version: cxapi.PROTO_RESPONSE_VERSION,
-      stacks: this.synthesizeStacks(Object.keys(this.stacks)),
-      runtime: this.collectRuntimeInformation()
-    };
+    // the three holy phases of synthesis: validate, prepare and synthesize
```

Comment:
Prepare should go before validate though. Don't want prepare to be able to mutate the tree into a state that is not valid.

Reply:
Good point